drivers/watchdog/wdt_gecko.c

```diff
@@ -0,0 +1,296 @@
+	 * timeout period up to 87.5% (= 7 * 12.5%)
+	 */
+	while (idx < 7) {
+		if (window > comp_val) {
+			comp_val += incr_val;
+			idx++;
+			continue;
+		}
+
+		break;
+	}
+
+	return idx;
+}
+
+static int wdt_gecko_setup(struct device *dev, u8_t options)
+{
+	const struct wdt_gecko_cfg *config = DEV_CFG(dev);
+	struct wdt_gecko_data *data = DEV_DATA(dev);
+	WDOG_TypeDef *wdog = config->base;
+
+	if (!data->timeout_valid) {
+		LOG_ERR("No valid timeouts installed");
+		return -EINVAL;
+	}
+
+	data->wdog_config.em2Run =
+		(options & WDT_OPT_PAUSE_IN_SLEEP) == 0U;
+	data->wdog_config.em3Run =
+		(options & WDT_OPT_PAUSE_IN_SLEEP) == 0U;
+
+	data->wdog_config.debugRun =
+		(options & WDT_OPT_PAUSE_HALTED_BY_DBG) == 0U;
+
+	WDOGn_IntClear(wdog, WDOG_IEN_TOUT);
+	if (data->callback != NULL) {
+		/* Interrupt mode for window */
+		/* Enable timeout interrupt */
+		WDOGn_IntEnable(wdog, WDOG_IEN_TOUT);
+	} else {
+		/* Disable timeout interrupt */
+		WDOGn_IntDisable(wdog, WDOG_IEN_TOUT);
+	}
+
+	/* Watchdog is started after initialization */
+	WDOGn_Init(wdog, &data->wdog_config);
+	LOG_DBG("Setup the watchdog");
+
+	return 0;
+}
+
+static int wdt_gecko_disable(struct device *dev)
+{
+	const struct wdt_gecko_cfg *config = DEV_CFG(dev);
+	struct wdt_gecko_data *data = DEV_DATA(dev);
+	WDOG_TypeDef *wdog = config->base;
+
+	WDOGn_Enable(wdog, false);
+	data->timeout_valid = false;
+	LOG_DBG("Disabled the watchdog");
+
+	return 0;
+}
+
+static int wdt_gecko_install_timeout(struct device *dev,
+				     const struct wdt_timeout_cfg *cfg)
+{
+	struct wdt_gecko_data *data = DEV_DATA(dev);
+	WDOG_Init_TypeDef init_defaults = WDOG_INIT_DEFAULT;
+	u32_t installed_timeout;
+
+	if (data->timeout_valid) {
+		LOG_ERR("No more timeouts can be installed");
+		return -ENOMEM;
+	}
+
+	if ((cfg->window.max < wdt_gecko_get_timeout_from_persel(0)) ||
+		(cfg->window.max > wdt_gecko_get_timeout_from_persel(
+			WDT_GECKO_MAX_PERIOD_SELECT_VALUE))) {
+		LOG_ERR("Upper limit timeout out of range");
+		return -EINVAL;
+	}
+
+	data->wdog_config = init_defaults;
+
+	data->wdog_config.perSel = (WDOG_PeriodSel_TypeDef)
+		wdt_gecko_get_persel_from_timeout(cfg->window.max);
+
+	installed_timeout = wdt_gecko_get_timeout_from_persel(
+		data->wdog_config.perSel);
+	LOG_INF("Installed timeout value: %u", installed_timeout);
+
+	if (cfg->window.min > 0) {
+		/* Window mode. Use rounded up timeout value to
+		 * calculate minimum window setting.
+		 */
+		data->wdog_config.winSel = (WDOG_WinSel_TypeDef)
+			wdt_gecko_convert_window(cfg->window.min,
+						installed_timeout);
+
+		LOG_INF("Installed window value: %u",
+			(installed_timeout / 8) * data->wdog_config.winSel);
+	} else {
+		/* Normal mode */
+		data->wdog_config.winSel = wdogIllegalWindowDisable;
+	}
+
+	/* Set mode of watchdog and callback */
+	switch (cfg->flags) {
+	case WDT_FLAG_RESET_SOC:
+	case WDT_FLAG_RESET_CPU_CORE:
+		data->wdog_config.resetDisable = false;
```

Comment:
#20693 suggests this platform does not provide a delay in which a callback can be invoked before the reset occurs.  If so I believe the documented API would require that this section return `-ENOTSUP` if a callback was passed.

Reply:
Thank you for insight! I changed the driver to reflect the -ENOTSUP status. Please let me know what you think.